frontend/src/pages/projects/projectSharing/ProjectSharingTable.tsx

```diff
@@ -0,0 +1,87 @@
+import * as React from 'react';
+import Table from '~/components/Table';
+import { RoleBindingKind } from '~/k8sTypes';
+import { deleteRoleBinding, generateRoleBindingProjectSharing, createRoleBinding } from '~/api';
+import ProjectSharingTableRow from './ProjectSharingTableRow';
+import { columnsProjectSharingUser, columnsProjectSharingGroup } from './data';
+import { ProjectSharingRBType } from './types';
+
+type ProjectSharingTableProps = {
+  type: ProjectSharingRBType;
+  permissions: RoleBindingKind[];
+  typeAhead: string[];
+  onCancel: () => void;
+  onError: (error: Error) => void;
+  refresh: () => void;
+};
+
+const ProjectSharingTable: React.FC<ProjectSharingTableProps> = ({
+  type,
+  permissions,
+  typeAhead,
+  onCancel,
+  onError,
+  refresh,
+}) => {
+  const [editCell, setEditCell] = React.useState<string[]>([]);
+
+  return (
+    <Table
+      variant="compact"
+      data={permissions}
+      columns={
+        type === ProjectSharingRBType.USER ? columnsProjectSharingUser : columnsProjectSharingGroup
+      }
+      disableRowRenderSupport
+      rowRenderer={(rb) => (
+        <ProjectSharingTableRow
+          key={rb.metadata?.name || ''}
+          obj={rb}
+          isEditing={rb.subjects[0]?.name === '' || editCell.includes(rb.metadata.name)}
+          typeAhead={typeAhead}
+          onCreateOrEditRoleBinding={(name, roleType, obj) => {
+            const newRBObject = generateRoleBindingProjectSharing(
+              obj.metadata.namespace,
+              type,
+              name,
+              roleType,
+            );
+            if (obj.subjects[0]?.name === '') {
+              createRoleBinding(newRBObject)
+                .then(() => refresh())
+                .catch((e) => {
+                  onError(e);
+                });
+            } else {
+              createRoleBinding(newRBObject)
+                .then(() =>
+                  deleteRoleBinding(obj.metadata.name, obj.metadata.namespace)
```

Comment:
Maybe a `Promise.all` dry run then run?

Reply:
Yeah, I thought of that, my fear was that the dry run completion wasn't deterministic and that could cause trouble, let me explain:

- `Delete old admin permission --> Create admin permission` Will fail since the user no longer has the permissions
- `Create admin permission --> Delete the old one` Will definitely success since the user will still have a role

I can test the dryRun solution, but my fear is that this might create a race condition. Thoughts on this?